### ISSUE TITLE
Use CloudCompare ICP in my project

### ISSUE THREAD

**Issue by hello-jo** [CloudCompare/CCCoreLib#81]
Hello, I am trying to make my interpretation of ICP algorithm to match 3d point clouds. But I found that ICP from Cloud Compare works great. I tried compiling the code in CCCoreLib, but had no idea how to read my own point cloud file. And I found that 2018 had the same questioner, but he used the <SimpleCloud.h> has been deleted by the new version. Is there any new way to read my own source point cloud and target point cloud?Looking forward to your reply. Thank you.

**Comment by dgirardeau** [comment]
The new point cloud class is the more generic 'CCCoreLib::PointCloud' now (https://github.com/CloudCompare/CCCoreLib/blob/master/include/PointCloud.h).

**Comment by hello-jo** [comment]
Thank you for your reply. Now I have successfully compiled the new version CCCoreLib. And I made simple program, where I am trying to create cloud based on.csv files. I tried to add the first point XYZ data into SC, but it was always unsuccessful and I had no idea. Here is my code. I wonder if you have any code suggestions for input point cloud.I would appreciate it. Thank you very much.

	double *original_high = new double[30 * 3];
	ifstream fin("test.csv");
	string line;
	int counter_csv = 0;
	while (getline(fin, line))
	{
		istringstream sin(line);
		string field;
		while (getline(sin, field, ','))
		{
			original_high[counter_csv] = atof(field.c_str());
			counter_csv++;
		}
	}
	std::vector<double> v;
	for (int i = 0; i < 90; ++i)
	{
		v.push_back(original_high[i]);
	}

	CCCoreLib::PointCloud *sc;
	sc->reserve(5);
	CCVector3 *vector = new CCVector3(v[0], v[1], v[2]);
	sc->addPoint(*vector);

**Comment by dgirardeau** [comment]
You have to instantiate the cloud ;)
`CCCoreLib::PointCloud* sc = new CCCoreLib::PointCloud("my cloud");`

And mind that you don't need to dynamically allocate your CCVector instances,it will create memory leaks. You just need to write:
```
CCVector3 vector(v[0], v[1], v[2]);
sc->addPoint(vector);
```

**Comment by hello-jo** [comment]
Thank you very much. Solved the problem I was facing.

**Comment by hello-jo** [comment]
Sorry for so much questions, but now I have successfully registered two point clouds. I cannot output the complete transformation matrix, so I output the transformation matrix one by one in the following way. I find that the result deviates greatly from the translation matrix calculated by CC software, and it seems that the result calculated by the software is correct.

**Comment by hello-jo** [comment]
Sorry for so much questions, but now I have successfully registered two point clouds. I cannot output the complete transformation matrix("*transformation"), so I output the transformation matrix one by one in the following way. I find that the result deviates greatly from the translation matrix calculated by CC software, and it seems that the result calculated by the software is correct. Can you tell me where am I wrong in this situation? I looked in RegistrationTools.h and all I can see is that default values of Parameters for ICP are setted in better way so I am not changed it. CCCoreLib::ICPRegistrationTools::ScaledTransformation *transofrmation = new CCCoreLib::ICPRegistrationTools::ScaledTransformation();
 double RMS = 0.00; 
 unsigned int pointcount = 0;
 CCCoreLib::ICPRegistrationTools::Parameters *params = new CCCoreLib::ICPRegistrationTools::Parameters();
 CCCoreLib::ICPRegistrationTools::Register(sc, 0, dst, *params, *transofrmation, RMS, pointcount);

 //Outputs the values in the R_matrix in order
 cout << "R_matrix:" << endl;
 for (int i = 0; i < 9; i++)
 {
  cout << *(*transofrmation->R.m_values + i) << endl;
 }

 //Outputs the values in the T_matrix in order
 cout << "T_matrix:" << endl;
 cout << transofrmation->T.x << endl << transofrmation->T.y << endl << transofrmation->T.z << endl;

**Comment by hello-jo** [comment]
Sorry for so much questions, but now I have successfully registered two point clouds. I cannot output the complete transformation matrix("*transformation"), so I output the transformation matrix one by one in the following way. I find that the result deviates greatly from the translation matrix calculated by CC software, and it seems that the result calculated by the software is correct.  I looked in RegistrationTools.h and all I can see is that default values of Parameters for ICP are setted in better way so I am not changed it. Can you tell me where am I wrong in this situation? Looking forward to your reply. Thank you.

 CCCoreLib::ICPRegistrationTools::ScaledTransformation *transofrmation = new CCCoreLib::ICPRegistrationTools::ScaledTransformation();
 double RMS = 0.00; 
 unsigned int pointcount = 0;
 CCCoreLib::ICPRegistrationTools::Parameters *params = new CCCoreLib::ICPRegistrationTools::Parameters();
 CCCoreLib::ICPRegistrationTools::Register(sc, 0, dst, *params, *transofrmation, RMS, pointcount);

 //Outputs the values in the R_matrix in order
 cout << "R_matrix:" << endl;
 for (int i = 0; i < 9; i++)
 {
  cout << *(*transofrmation->R.m_values + i) << endl;
 }

 //Outputs the values in the T_matrix in order
 cout << "T_matrix:" << endl;
 cout << transofrmation->T.x << endl << transofrmation->T.y << endl << transofrmation->T.z << endl;

**Comment by hello-jo** [comment]
Thank you for your reply. But I found a problem in CC software version _2.12.4_. I don't know if anyone has asked this question, and I haven't seen it in Issues. First I made a known matrix 'T' change to the source point cloud and saved it. Second, I open the saved point cloud and find that it is in the position where the matrix 'T' has been transformed twice. However, I did not find this problem in CC software version _![3576b53b4e9750e1241bce45415dba5](https://user-images.githubusercontent.com/82424028/186608527-eb4b6b60-daa4-40f2-a223-25a1efe13381.png)_. This led to my discovery yesterday that the calculated results of the compiled code were different from those of the CC software.

**Comment by hello-jo** [comment]
Thank you for your reply. But I found a problem in CC software version _2.12.4_. I don't know if anyone has asked this question, and I haven't seen it in Issues. First I made a known matrix 'T' change to the source point cloud and saved it. Second, I open the saved point cloud and find that it is in the position where the matrix 'T' has been transformed twice. However, I did not find this problem in CC software version _2.11.3_. This led to my discovery yesterday that the calculated results of the compiled code were different from those of the CC software.
![3576b53b4e9750e1241bce45415dba5](https://user-images.githubusercontent.com/82424028/186609178-e693c18f-678a-40a1-b464-9dab7f515dbc.png)
